---
title: 'Analysis: NNK7 repressed set'
output:
  pdf_document: default
  html_document: default
date: "2023-01-19"
---

```{r setup, include=FALSE}
# Gets thing set up. Need the library, and some functions.

knitr::opts_chunk$set(echo = TRUE, comment = '', fig.width = 6, fig.height = 6)

library(tidyverse)
library(ggplot2)

source("C:/Users/worms/Dropbox/PhD/PhD-Scripts/CyclicPeptidePipeline/function.R")
```

# Introduction

This is the analysis of the induced NNK7 peptides from the drift experiment. The required datasets are created in the "peptideAnalysis.R" file. I'm going to do it on a subset of 100 000 reads at first to speed things up. This report looks at the subset that is repressed.


```{r 'load datasets', echo=FALSE}
directory <- "C:/Users/worms/NGS Data/2022.06.07_drift_seq/90-666155004b/00_fastq/NNK/NNK7"
repressed_set <- read.csv(file = paste(directory, "repressed_set_NNK7.csv", sep = "/"),
                        nrow = 100000) 

```

## Summary

Let's have a look at the distribution of enrichment ration:

```{r echo=TRUE}
summary(repressed_set$enrichment_ratio)
```
We can see that in induced conditions, most peptides becomes slightly less common. None are strongly purged and some are highly enriched. There are however a number of peptides with relatively low (< 0.6). That's `r repressed_set %>% filter(enrichment_ratio < 0.6) %>% nrow()`, or `r (repressed_set %>% filter(enrichment_ratio < 0.6) %>% nrow()/nrow(repressed_set))*100` % of the total. I take the large number of peptides with an enrichment ratio of <1 to basically be neutrally selected. Overall out of `r nrow(repressed_set)` genes encoding peptides, `r repressed_set %>% filter(enrichment_ratio < 1) %>% nrow()` have a enrichment ratio of less than 1, `r repressed_set %>% filter(between(enrichment_ratio,1,2)) %>% nrow()` are weakly selected (enrichment ratio of between 1 and 2) and `r repressed_set %>% filter(enrichment_ratio > 2) %>% nrow()` (`r repressed_set %>% filter(enrichment_ratio > 2) %>% nrow() / nrow(repressed_set) * 100`%) have an enrichment ratio of over 2.

```{r, 'graph low enrichment', echo=FALSE}
ggplot(repressed_set %>%
         filter(enrichment_ratio < 1.25), aes(enrichment_ratio)) +
  geom_histogram(bins = 30, color = "black", fill = "lightblue") +
  labs(x = "Enrichment Ratio", y = "Frequency") +
  ggtitle("Histogram of the enrichment ratio for enrichment < 1.25")
```
```{r echo=FALSE}
get_enrichment_ratio_sd <- function(data.frame, max_ratio = 2, round_sd = FALSE){

    x <- data.frame %>%
filter(enrichment_ratio < max_ratio) %>%
select(enrichment_ratio) %>%
.[[1]] %>%
sd()
```



Compared to the induced set, the distribution of enrichment ratio seems a bit wider. This seems to old if you look at the standard deviation of weakly enriched genes for the induced set (`r induced_set %>% get_enrichment_ratio_sd(max_ratio = 1.5, round_sd = TRUE)`) and repressed set (`r repressed_set %>% get_enrichment_ratio_sd(max_ratio = 1.5, round_sd = TRUE)`). That's interesting, but I don't know if it is meaningful. 

```{r}

  
  if(round_sd){
    return(x %>% round(digits = 2))
  }
  else{
    return(x)
  }
  
}

induced_set %>% get_enrichment_ratio_sd(max_ratio = 1.5)
repressed_set %>% get_enrichment_ratio_sd(max_ratio = 1.5)
```


If we look at it taking the log2 of the enrichment ratio, we get the followin graph. There seems to be a relatively uniform distribution of genes with enrichment ratio between 2 and 128.

```{r echo=FALSE}
ggplot(repressed_set %>%
         filter(enrichment_ratio_log > 1), aes(enrichment_ratio_log)) +
  geom_histogram(bins = 30, color = "black", fill = "lightblue") +
  labs(x = "Log(Enrichment Ratio)", y = "Frequency") +
  ggtitle("Histogram of the enrichment ratio for log(enrichment) > 1")
```


This is strange as we expect some toxic peptides. And they don't seem to be many peptides that are purged!

## Peptides with stop codons

```{r echo=FALSE}
repressed_set_stops <- repressed_set %>%
  filter(grepl("\\*", peptide_seq))
```

I'll now subset the `r nrow(repressed_set_stops)` sequences (`r (nrow(repressed_set_stops)/nrow(repressed_set)*100) %>% round(digits = 2)`% of the total) that contains stop codons. The percentage I would expect with peptides containing 7 randomized NNK codons is `r ((1-(31/32)^7)*100) %>% round(digits = 2)`% of the total, so this is consistent. 

```{r echo=FALSE}
ggplot(repressed_set_stops %>%
         filter(enrichment_ratio < 1.25), aes(enrichment_ratio)) +
  geom_histogram(bins = 30, color = "black", fill = "lightblue") +
  labs(x = "Enrichment Ratio", y = "Frequency") +
  ggtitle("Histogram of the enrichment ratio of peptides with stop codons for enrichment < 1.25")
```


```{r echo=FALSE}
ggplot(repressed_set_stops %>%
         filter(enrichment_ratio_log > 0.5), aes(enrichment_ratio_log)) +
  geom_histogram(bins = 10, color = "black", fill = "lightblue") +
  labs(x = "Log(Enrichment Ratio)", y = "Frequency") +
  ggtitle("Histogram of the enrichment ratio of peptides with stop codons for log(enrichment) > 0.5")
```

The general patterns of stop codon containing peptides is similar to the overall pattern, which kind of sucks. `r repressed_set_stops %>% filter(enrichment_ratio > 2) %>% nrow()` (`r (repressed_set_stops %>% filter(enrichment_ratio > 2) %>% nrow() / nrow(repressed_set_stops) * 100) %>% round(digits = 2)`%) have an enrichment ratio of over 2. This is almost exactly the same as those without stop codons. 




```{r echo=FALSE}

```




