---
title: "Analysis of short cyclic peptides"
author: "Sebastian Worms"
date: "`r format(Sys.time(), '%B %d, %Y')`"
output:
  html_document:
    toc: true # table of content true
    toc_depth: 3  # upto three depths of headings (specified by #, ## and ###)
    number_sections: true  ## if you want number sections at each table header
    theme: united  # many options for theme, this one is my favorite.
    highlight: tango  # specifies the syntax highlighting style
    df_print: paged
  html_notebook: default
  pdf_document: default
---

```{r setup, include=FALSE, echo=FALSE}
knitr::opts_chunk$set(echo = FALSE, comment = '', fig.width = 6, fig.height = 6)

library(tidyverse)
library(ggplot2)

#Load some custom functions
source("../functions/function_analysis.R")

#Work with a limited number of reads to get something I can work with easily.
n_reads <- 1*10^6

```

# Introduction

This is the analysis of the peptides from the drift experiment. The required datasets were created in the "peptide_analysis.R" file. I'm going to do it on a subset of of `r print(n_reads)` reads for the short dataset for now.

```{r 'load datasets', echo=FALSE}
directory <- "C:/Users/worms/ngs_data/2022_06_07_drift_seq/90-666155004b/00_fastq/all_files"

count_set <- read.csv2(file = file.path(directory, "count_set_short.csv"),
                  nrow = n_reads) 

# Going to drop all the infinite enrichment ratio to NA
count_set <- count_set %>%
  mutate(enrichment_ratio = ifelse(is.infinite(enrichment_ratio),NA,enrichment_ratio)) %>%
  mutate(enrichment_ratio_log = ifelse(is.infinite(enrichment_ratio_log),NA,enrichment_ratio_log)) 
```

## Read quality check

### Stop codons

```{r}
unique_sequences <- count_set %>% pull(peptide_seq) %>% unique() %>% length()
sequences_with_stop <- count_set %>% filter(grepl("\\*", peptide_seq)) %>% pull(peptide_seq) %>% unique() %>% length()
percent_with_stop <- sequences_with_stop/unique_sequences

```

Let's look at the subset of sequences with stops codons. `r sequences_with_stop` sequences (`r percent_with_stop`)% of the total) contains stop codons. A lot of those are lower quality, not starting with TGC.

### Reads not starting with TGC

```{r `subsetting`}
count_set_good <- count_set %>%
    filter(grepl("^TGC.*", seq))

unique_sequences_good <- count_set_good %>% pull(peptide_seq) %>% unique() %>% length()
coverage <- (unique_sequences_good/21^3*100) %>% round(digits = 2)

average_good_reads <- count_set_good %>% select(c("gen1_12", "gen5_12")) %>% rowSums() %>% mean() %>% round(digits = 0)
average_bad_reads <- count_set %>% filter(!grepl("^TGC.*", seq)) %>% select(c("gen1_12", "gen5_12")) %>% rowSums() %>% mean() %>% round(digits = 0)

rm(count_set)
```

I subtract the sequences that do not start with the correct codon (*TGC*) and work only with those for the rest of this document. Out of `r unique_sequences`, this leaves me `r unique_sequences_good`. 

The sequences not starting with TGC are mostly lower quality reads. The average number of reads (over all library and generation) per sequence is `r average_good_reads` for the sequences starting with TGC, but just `r average_bad_reads` for those not starting with TGC. 

```{r}
count_set <- count_set_good
rm(count_set_good)
```


### Coverage

```{r 'checking coverage'}
dna_unique <- count_set %>% pull("seq") %>% unique() %>% length()

```

I can expect at most `r 21^3` different peptide sequences (counting stops), so my coverage is `r coverage`%. Pretty good. At the DNA level, I expect `r 32^3` unique sequences with NNK codons and I have `r dna_unique`. This is presumably because I have low-quality sequence where the third base of my codons isn't a G or T. I will filter those too. 

```{r 'filtering NNK'}
sequences_after_nnk_filtering <- count_set %>% 
  filter(grepl("^TGC([ATGC][ATGC][GT]){3}$", seq)) %>% 
  pull(seq) %>% 
  unique() %>% 
  length()

sequences_filtered_by_nnk_filtering <- dna_unique - sequences_after_nnk_filtering
 
average_reads_after_nnk_filtering <- count_set %>% filter(grepl("^TGC([ATGC][ATGC][GT]){3}$", seq)) %>%  select(c("gen1_12", "gen5_12")) %>% rowSums() %>% mean() %>% round(digits = 0)

average_reads_nnk_filtered_seq <- count_set %>% filter(!grepl("^TGC([ATGC][ATGC][GT]){3}$", seq)) %>%  select(c("gen1_12", "gen5_12")) %>% rowSums() %>% mean() %>% round(digits = 0)

coverage_after_nnk_filtering <- ((sequences_after_nnk_filtering/(32^3))*100) %>% round(digits = 2)

count_set <- count_set %>%
    filter(grepl("^TGC([ATGC][ATGC][GT]){3}$", seq))

```

This filtering step removes `r sequences_filtered_by_nnk_filtering` sequences with an average of `r average_reads_nnk_filtered_seq` reads each. `r sequences_after_nnk_filtering` sequences remains with an average of `r average_reads_after_nnk_filtering` each. This is `r coverage_after_nnk_filtering`% coverage at the DNA level.

### Coverage and abundance in gen 1

I'm going to check specifically the gen 1 coverage and distribution of abundance pre-selection. One concern is that the library was really skewed, and that some sequences were missing or almost missing.

```{r}
unique_gen1_sequence <- count_set %>%
  filter(gen1_12 != 0) %>%
  pull(seq) %>%
  unique() %>%
  length()
```

So `r unique_gen1_sequence` are present in generation 1, that most of them (remember `r sequences_after_nnk_filtering` total sequences.).

```{r}
p <- ggplot(data = count_set %>% mutate(gen1_12), aes(x = library, y = gen1_12)) +
  geom_violin() + 
  labs(x = "Library", y = "Count") +
  ylim(0,100)

p %>% print() %>% suppressWarnings()
```

So most have a count in the single or low two digits, but there is a very, very long tail. (Note that the above graph was truncated. The tail goes on to the tens of thousands.)

## Enrichment ratio

Let's have a look at the distribution of enrichment ratio, having dropped all the enrichment ratios that were infinite, or zeroes (so not present in one of the two generation). Overall it looks like a pretty symmetrical set in the NNB library, but there is a lot of heavily purged sequences in the induced section of NNK.

```{r}
tapply(count_set %>% filter(enrichment_ratio > 0) %>% pull(enrichment_ratio), 
       count_set %>% filter(enrichment_ratio > 0) %>% mutate(lib_condition = paste(library, induction)) %>% pull(lib_condition), 
       summary)

```

```{r 'histogram enrichment'}
p <- ggplot(count_set, aes(x = enrichment_ratio_log)) +
  geom_histogram(color = "black", fill = "white") +
  labs(title = "Histogram of log(Enrichment Ratio", x = "Log2(Enrichment Ratio)", y = "Count") +
  facet_grid(library~induction)

p %>% print() %>% suppressWarnings()

```
The induced numers for NNK are interesting, why doesn't it repeat for the NNB library? Could it be that those are mixed too? :/

```{r}
get_sequence_number <- function(df){
  count <- df %>%
    unique() %>%
    length()
  return(count)
}

tapply(count_set %>% pull(seq), 
       count_set %>% mutate(lib_condition = paste(library, induction)) %>% pull(lib_condition), 
       get_sequence_number) %>% as.matrix() %>% kable()
```

The NNB libraries are larger than the NNK who only have around 2/3 of total coverage.

The very high number could be PCR artifact? If so they shouldn't correlate between generation.

```{r 'Test correlation'}
ggplot(data = count_set, aes(x= gen1_12, y = gen5_12)) +
  geom_point() +
  facet_grid(library~induction)
```

Okay, this looks decent. I'm started to believe those numbers. A bit worried at the small number of sequences in the NNK set.

### Purged count_set sequences

```{r}

```




#### Purge

No sequences seems missing or highly repressed. So no strong purge. What could have happened is that some of the ones in the induced set that are missing are actually very strongly purged there compared to the repressed set. If true, I expect them to be purged in the repressed set too.

```{r 'Testing the hypothesis that the missing peptides are purged'}

```

The absent peptides are all containing relatively rare codons, with a lot of phenylalanine, tryptophane, cysteines... That could be the explanation. Looking at the data sets, all eleven peptides are "NAs" in the represed condition too, so relatively rare.

### Correlation of Enrichment ratios

I should expect the enrichment ratio for the same peptide sequence to correlate pretty well.

```{r 'Looking at the correlation of enrichment_ratio'}

```

There is fairly good correlation (r correlation %>% round(digit = 2)) between the induced and repressed condition. The plot isn't fully symmetrical which I don't fully understand.

```{r, include=TRUE}

```

### 'RR' motifs

RR motifs are known to interfere with the membrane. Let look at the peptides with those motifs.

```{r}
count_set_RR <- count_set %>% 
      filter(grepl("RR", peptide_seq))

tapply(count_set$enrichment_ratio, count_set$induction, summary)

# Color the points with RR motifs

# mean_enrichment_ratio_log$RR <- mean_enrichment_ratio_log$peptide_seq %>% grepl("RR", .) %>% ifelse(., "RR", "not RR")
# 
# p <- ggplot(mean_enrichment_ratio_log, aes(x = enrichment_ratio_induced, y = enrichment_ratio_repressed, color = RR)) +
#   geom_point(alpha = 0.5, size = 1) +
#   facet_grid(RR ~ .) +
#   xlab("Log2(Enrichment ratio) in the induced condition") +
#   ylab("Log2(Enrichment ratio) in the repressed condition")


```

### Enrichment by charge

### Background

### Peptides with Proline

### Hydrophobicity

### Initial ration and enrichment ratios

### His-rich peptides
