---
title: "Analysis of short cyclic peptides"
author: "Sebastian Worms"
date: "`r format(Sys.time(), '%B %d, %Y')`"
output:
  html_document:
    toc: true # table of content true
    toc_depth: 3  # upto three depths of headings (specified by #, ## and ###)
    number_sections: true  ## if you want number sections at each table header
    theme: united  # many options for theme, this one is my favorite.
    highlight: tango  # specifies the syntax highlighting style
    df_print: paged
  html_notebook: default
  pdf_document: default
---

```{r setup, include=FALSE, echo=FALSE}
knitr::opts_chunk$set(echo = FALSE, comment = '', fig.width = 6, fig.height = 6)

library(tidyverse)
library(ggplot2)

#Load some custom functions
source("../functions/function_analysis.R")

#Work with a limited number of reads to get something I can work with easily.
n_reads <- 1*10^6

```

# Introduction

This is the analysis of the peptides from the drift experiment. The required datasets were created in the "peptide_analysis.R" file. I'm going to do it on a subset of of `r print(n_reads)` reads for the short dataset for now.

```{r 'load datasets', echo=FALSE}
directory <- "C:/Users/worms/ngs_data/2022_06_07_drift_seq/90-666155004b/00_fastq/all_files"

count_set <- read.csv2(file = file.path(directory, "count_set_short.csv"),
                  nrow = n_reads) 

# Going to drop all the infinite enrichment ratio to NA
count_set <- count_set %>%
  mutate(enrichment_ratio = ifelse(is.infinite(enrichment_ratio),NA,enrichment_ratio)) %>%
  mutate(enrichment_ratio_log = ifelse(is.infinite(enrichment_ratio_log),NA,enrichment_ratio_log))
```

## Short peptides

Let's have a look at the distribution of enrichment ratio:

```{r}
tapply(count_set$enrichment_ratio, count_set$induction, summary)


na_percent_induced <- count_set%>% 
  filter(induction == "induced") %>%
  select(enrichment_ratio) %>% 
  is.na() %>% 
  mean() %>%
  round(digits = 2)

na_percent_repressed <- count_set%>% 
  filter(induction == "repressed") %>%
  select(enrichment_ratio) %>% 
  is.na() %>% 
  mean() %>%
  round(digits = 2)
```

### Stop codons

```{r}
count_set_stops <- count_set %>%
  filter(grepl("\\*", peptide_seq))

count_set_no_stops <- count_set %>%
  filter(!grepl("\\*", peptide_seq))
```

Let's look at the subset of sequences with stops codons. `r nrow(count_set_stops)` sequences (`r (nrow(count_set_stops)/nrow(count_set)*100) %>% round(digits = 2)`% of the total) contains stop codons. The percentage I would expect with peptides containing 3 randomized NNK codons is `r ((1-(31/32)^3)*100) %>% round(digits = 2)`% of the total, so this is much higher than expected.Looking at the sequence I have something odd in that many of the sequences don't have a starting cysteine.

I will subtract the sequences that do not start with the correct codon (*TGC*).

```{r 'correct TGC', echo=FALSE}
count_set_correct <- count_set%>%
    filter(grepl("^TGC.*", seq))

count_set_stops_correct <- count_set_correct %>%
  filter(grepl("\\*", peptide_seq))

count_set_correct_no_stops <- count_set_correct %>%
  filter(!grepl("\\*", peptide_seq))
```

This is a surprisingly large number: `r nrow(count_set)-nrow(count_set_correct)` out of `r nrow(count_set)` (`r 100-((nrow(count_set_correct)/nrow(count_set)*100)%>% round(digit = 2))`%). Although that's the number of sequences, not the number or reads. We have `r count_set_correct %>% select(c("gen1_12","gen5_12")) %>% sum()` correct reads out of `r count_set %>% select(c("gen1_12","gen5_12")) %>% sum()` (`r ((count_set_correct %>% select(c("gen1_12","gen5_12")) %>% sum())/(count_set %>% select(c("gen1_12","gen5_12")) %>% sum()) *100) %>% round(digit = 2)`%), almost all of them.

So looking at only the sequences starting with TGC, we have `r (nrow(count_set_stops_correct)/nrow(count_set_correct)*100) %>% round(digits = 2)`% of stop codons, much closer to the expected `r ((1-(31/32)^3)*100) %>% round(digits = 2)`%. We'll subset only the correct ones for downstream analysis and work with the count_set_correct data set.

### NAs

NAs here are the cases where there were no reads at the generation 1 but some after five generations. `r na_percent_induced * 100`% of genes in the induced set and `r na_percent_repressed * 100`% in the repressed sets are such NAs. Interesting that way more reads are such NAs for the repressed set, despite it being generally less enriched.

Compared to the rest of the counts, the NAs usually have very low sequence counts even at the generation five and so probably represent rare peptides. That would explain the higher numbers of NAs in the repressed set.

```{r 'select NAs'}
count_set_na <- count_set[count_set%>% is.na() %>% rowSums() != 0,]
tapply(count_set_na$gen5_12, count_set_na$induction, summary)
tapply(count_set$gen5_12, count_set$induction, summary)

```

If those are rare, "fake" sequences that are artifact, I expect a very high rate of sequences that don't start with *TGC*. Let's check that.

```{r 'correct sequence NA'}
count_set_na_correct_nombre <- count_set_na %>%
    filter(grepl("^TGC.*", seq)) %>%
    nrow()
```

Out of `r nrow(count_set_na)` reads with NAs, `r count_set_na_correct_nombre` have a sequence that starts with *TGC*. That's `r 100-((count_set_na_correct_nombre/nrow(count_set_na)*100)%>% round(digit = 2))`% of wrong sequences, twice the average rate! This fits with the idea that rare sequences are the one causing issues.

### Purged count_set sequences

```{r}
count_set_correct_no_na <- na.omit(count_set_correct)
count_set_correct %>%
  filter(enrichment_ratio < 1) %>%
  select(enrichment_ratio) %>%
  summary
```

Most peptides here are weakly purged. Of the `r nrow(count_set_correct_no_na)` peptides with an enrichment ratio that was computed, we have `r count_set_correct_no_na %>% filter(enrichment_ratio < 1) %>% nrow()` that are purged. But only `r count_set_correct_no_na %>% filter(enrichment_ratio < 0.5) %>% nrow()` (`r ((count_set_correct_no_na$enrichment_ratio %>% between(.,0,0.5) %>% mean())*100 )%>% round(.,digit = 2)`%) have seen their frequencies reduced at least twofold.

#### What is the coverage?

```{r 'checking coverage'}
number_peptide_repressed <- count_set_correct %>%
  filter(!grepl("/*", peptide_seq)) %>%
  filter(induction == "repressed") %>%
  select(peptide_seq) %>%
  unique() %>%
  nrow()

number_peptide_induced <- count_set_correct %>%
  filter(!grepl("/*", peptide_seq)) %>%
  filter(induction == "induced") %>%
  select(peptide_seq) %>%
  unique() %>%
  nrow()

number_peptide_induced_no_na <- count_set_correct %>%
  na.omit() %>%
  filter(!grepl("/*", peptide_seq)) %>%
  filter(induction == "induced") %>%
  select(peptide_seq) %>%
  unique() %>%
  nrow()

number_peptide_repressed_no_na <- count_set_correct %>%
  na.omit() %>%
  filter(!grepl("/*", peptide_seq)) %>%
  filter(induction == "repressed") %>%
  select(peptide_seq) %>%
  unique() %>%
  nrow()
```

With count_set peptides, we expect to have 20\^3 or `r 20^3` different possible peptide sequences (not counting cyclisation). In my case, I have `r number_peptide_repressed` peptides in the repressed condition and `r number_peptide_induced` in the induced condition, so almost perfect to perfect coverage. Those numbers are `r number_peptide_repressed_no_na` and `r number_peptide_induced_no_na` if one removes NAs.

#### Purge

No sequences seems missing or highly repressed. So no strong purge. What could have happened is that some of the ones in the induced set that are missing are actually very strongly purged there compared to the repressed set. If true, I expect them to be purged in the repressed set too.

```{r 'Testing the hypothesis that the missing peptides are purged'}

induced_set_peptide_sequence <- count_set_correct %>% #Extract all the peptide sequence in the induced set
  filter(!grepl("/*", peptide_seq)) %>% 
  filter(induction == "induced") %>%
  select(peptide_seq) %>%
  unique() 

repressed_set_peptide_sequence <- count_set_correct %>% #Extract all the peptide sequence in the repressed set
  filter(!grepl("/*", peptide_seq)) %>% 
  filter(induction == "repressed") %>%
  select(peptide_seq) %>%
  unique()

absent_peptides <- repressed_set_peptide_sequence %>% filter( !(repressed_set_peptide_sequence$peptide_seq %in% induced_set_peptide_sequence$peptide_seq))

absent_peptides

count_set_absent_peptides <- count_set %>%
  filter(count_set$peptide_seq %in% absent_peptides$peptide_seq)
```

The absent peptides are all containing relatively rare codons, with a lot of phenylalanine, tryptophane, cysteines... That could be the explanation. Looking at the data sets, all eleven peptides are "NAs" in the represed condition too, so relatively rare.

### Correlation of Enrichment ratios

I should expect the enrichment ratio for the same peptide sequence to correlate pretty well.

```{r 'Looking at the correlation of enrichment_ratio'}
count_set_correct_no_na_no_stops <- count_set_correct_no_na %>% 
  filter(!grepl("/*", peptide_seq))

mean_enrichment_ratio_log_repressed <-  count_set_correct_no_na_no_stops %>%
  subset(induction == "repressed", 
         select=c(peptide_seq, enrichment_ratio_log)) %>%
  aggregate(enrichment_ratio_log ~ peptide_seq, data=., FUN=mean)
  
mean_enrichment_ratio_log_induced <-  count_set_correct_no_na_no_stops %>%
  subset(induction == "induced", 
         select=c(peptide_seq, enrichment_ratio_log)) %>%
  aggregate(enrichment_ratio_log ~ peptide_seq, data=., FUN=mean)
 
mean_enrichment_ratio_log <- inner_join(mean_enrichment_ratio_log_repressed, mean_enrichment_ratio_log_induced, by = "peptide_seq")

colnames(mean_enrichment_ratio_log) <- c("peptide_seq", "enrichment_ratio_repressed", "enrichment_ratio_induced")

correlation <- cor(mean_enrichment_ratio_log$enrichment_ratio_repressed, mean_enrichment_ratio_log$enrichment_ratio_induced)

p <- ggplot(mean_enrichment_ratio_log %>% filter(enrichment_ratio_repressed > -1.5), aes(x = enrichment_ratio_induced, y = enrichment_ratio_repressed)) +
  geom_point(alpha = 0.1) +
  xlab("Log2(Enrichment ratio) in the induced condition") +
  ylab("Log2(Enrichment ratio) in the repressed condition") +
  geom_abline(slope = 1, intercept =0)

p

```

There is fairly good correlation (`r correlation %>% round(digit = 2)`) between the induced and repressed condition. The plot isn't fully symmetrical which I don't fully understand.

```{r, include=TRUE}
p
```

### 'RR' motifs

RR motifs are known to interfere with the membrane. Let look at the peptides with those motifs.

```{r}
count_set_RR <- count_set_correct_no_stops %>% 
      filter(grepl("RR", peptide_seq))

tapply(count_set$enrichment_ratio, count_set$induction, summary)

# Color the points with RR motifs

mean_enrichment_ratio_log$RR <- mean_enrichment_ratio_log$peptide_seq %>% grepl("RR", .) %>% ifelse(., "RR", "not RR")

p <- ggplot(mean_enrichment_ratio_log, aes(x = enrichment_ratio_induced, y = enrichment_ratio_repressed, color = RR)) +
  geom_point(alpha = 0.5, size = 1) +
  facet_grid(RR ~ .) +
  xlab("Log2(Enrichment ratio) in the induced condition") +
  ylab("Log2(Enrichment ratio) in the repressed condition")

p
```

### Enrichment by charge

### Background

### Peptides with Proline

### Hydrophobicity

### Initial ration and enrichment ratios

### His-rich peptides
